# Chapter 1: Exploratory Data Analysis

The question:

Do first time babies tend to arrive late?

Many anecdotal evidence because they are based on data that is unpublished and usually personal. Which fails because:

- **Small number of observations**: If pregnancy length is longer for first babies, the difference is probably small compared to natural variation. In that case, we might have to compare a large number of pregnancies to be sure that a difference exists.
- **Selection bias**: People who join a discussion of this question might be interested because their first babies were late. In that case the process of selecting data would bias the results.
- **Confirmation bias**: People who believe the claim might be more likely to contribute examples that confirm it. People who doubt the claim are more likely to cite counterexamples.
- **Inaccuracy**: Anecdotes are often personal stories, and often misremembered, misrepresented, repeated inaccurately, etc.

## Statistical Approach

To address the limitations of anecdotes, we will use the tools of statistics, which include:
Data collection: We will use data from a large national survey that
was designed explicitly with the goal of generating statistically valid
inferences about the U.S. population.
- **Descriptive statistics**: We will generate statistics that summarize the data concisely, and evaluate different ways to visualize data.
- **Exploratory data analysis**: We will look for patterns, differences, and other features that address the questions we are interested in. At the same time we will check for inconsistencies and identify limitations.
- **Estimation**: We will use data from a sample to estimate characteristics of the general population.
- **Hypothesis testing**: Where we see apparent effects, like a difference between two groups, we will evaluate whether the eect might have happened by chance.

## The Data Source

We will be usin the National Survey of Family Growth.

See [link](http://cdc.gov/nchs/nsfg.htm) and explore the different data sets and informarion.

The NSFG is a **cross-sectional** study, which means that it captures a snapshot of a group at a point in time. The most common alternative is a **longitudinal study**, which observes a group repeatedly over a period of time.

The goal of the survey is to draw conclusions about a **population**; the target population of the NSFG is people in the United States aged 15-44. Ideally surveys would collect data from every member of the population, but that's seldom possible. Instead we collect data from a subset of the population called a **sample**. The people who participate in a survey are called **respondents**.

In general, cross-sectional studies are meant to be **representative**, which means that every member of the target population has an equal chance of participating. That ideal is hard to achieve in practice, but people who conduct surveys come as close as they can.

The NSFG is not representative; instead it is deliberately **oversampled**. The designers of the study recruited three groups|Hispanics, African-Americans and teenagers|at rates higher than their representation in the U.S. population, in order to make sure that the number of respondents in each of these groups is large enough to draw valid statistical inferences.

Of course, the drawback of oversampling is that it is not as easy to draw conclusions about the general population based on statistics from the survey. We will come back to this point later.

The codebook and user's guide for the NSFG data are available from [link](http://www.cdc.gov/nchs/nsfg/nsfgcycle6.htm)

We will be usin the National Survey of Family Growth.

See [link](http://cdc.gov/nchs/nsfg.htm) and explore the different data sets and informarion.

The NSFG is a **cross-sectional study**, which means that it captures a snapshot of a group at a point in time. The most common alternative is a **longitudinal study**, which observes a group repeatedly over a period of time.

The goal of the survey is to draw conclusions about a **population**; the target population of the NSFG is people in the United States aged 15-44. Ideally surveys would collect data from every member of the population, but that's seldom possible. Instead we collect data from a subset of the population called a **sample**. The people who participate in a survey are called **respondents**.

In general, cross-sectional studies are meant to be **representative**, which means that every member of the target population has an equal chance of participating. That ideal is hard to achieve in practice, but people who conduct surveys come as close as they can.

The NSFG is not representative; instead it is deliberately **oversampled**. The designers of the study recruited three groups|Hispanics, African-Americans and teenagers|at rates higher than their representation in the U.S. population, in order to make sure that the number of respondents in each of these groups is large enough to draw valid statistical inferences.

Of course, the drawback of oversampling is that it is not as easy to draw conclusions about the general population based on statistics from the survey. We will come back to this point later.

The codebook and user's guide for the NSFG data are available from http://www.cdc.gov/nchs/nsfg/nsfgcycle6.htm

### Importing the data

Once you are done go to the folder ThinkStats2/code on your terminal and run nsfg.py:
> cd ThinkStats2/code

> python nsfg.py

You should get a message like 'All tests passed'

Now explore the data on the folder. How does 2002FemPreg.dct look?

This is a Stata dictionary file.

thinkstats2.py has a module to open Stata dictionaries.

A **module** is a Python object with arbitrarily named attributes that you can bind and reference. Simply, a module is a file consisting of Python code. A module can define functions, classes and variables. A module can also include runnable code.

Explore the module nsfg file in the folder. find the function ReadFemPreg() and then import it.

You might have to copy the module to the correct directory.

In [1]:
import pandas as pd
from Thinkstats2 import nsfg
import numpy as np

In [25]:
preg = nsfg.ReadFemPreg(dct_file='../Think_Stats/Thinkstats2/2002FemPreg.dct',
                      dat_file='Thinkstats2/2002FemPreg.dat.gz')

What are the columns in the Dataframe?

In [4]:
preg.columns

Index(['caseid', 'pregordr', 'howpreg_n', 'howpreg_p', 'moscurrp', 'nowprgdk',
       'pregend1', 'pregend2', 'nbrnaliv', 'multbrth',
       ...
       'laborfor_i', 'religion_i', 'metro_i', 'basewgt', 'adj_mod_basewgt',
       'finalwgt', 'secu_p', 'sest', 'cmintvw', 'totalwgt_lb'],
      dtype='object', length=244)

In [5]:
type(preg.columns)

pandas.core.indexes.base.Index

How many columns does it have? Use 2 methods to calculate it.

In [6]:
preg.shape[1]

244

In [7]:
len(preg.columns)

244

Remember that columns is not a **method** is an **attribute**

What is the first column?

In [8]:
preg.columns[0]

'caseid'

Access pregordr column. Use 2 different methods.

In [9]:
preg.loc[:,'pregordr'].head()

0    1
1    2
2    1
3    2
4    3
Name: pregordr, dtype: int64

In [10]:
preg.pregordr.head()

0    1
1    2
2    1
3    2
4    3
Name: pregordr, dtype: int64

What type is that column? and what type is the column object?

In [11]:
preg.pregordr.dtype

dtype('int64')

In [12]:
type(preg.pregordr)

pandas.core.series.Series

In [13]:
type(preg)

pandas.core.frame.DataFrame

Get the rows 2 to 4 of the column.

In [14]:
preg.pregordr[1:4]

1    2
2    1
3    2
Name: pregordr, dtype: int64

### Variables

Out of the 244 we are only going to use:

- *prglngth* is the integer duration of the pregnancy in weeks.
- *outcome* is an integer code for the outcome of the pregnancy. The code 1 indicates a live birth.
- *pregordr* is a pregnancy serial number; for example, the code for a respondent’s first pregnancy is 1, for the second pregnancy is 2, and so on.
- *birthord* is a serial number for live births; the code for a respondent’s first child is 1, and so on. For outcomes other than live birth, this field is blank.
- *birthwgt_lb* and *birthwgt_oz* contain the pounds and ounces parts of the birth weight of the baby.
- *agepreg* is the mother’s age at the end of the pregnancy.
- *finalwgt* is the statistical weight associated with the respondent. It is a floating-point value that indicates the number of people in the U.S. population this respondent represents.

If you read the codebook carefully, you will see that many of the variables are **recodes**, which means that they are not part of the raw data collected by the survey; they are calculated using the **raw data**.

For example, prglngth for live births is equal to the raw variable wksgest (weeks of gestation) if it is available; otherwise it is estimated using mosgest * 4.33 (months of gestation times the average number of weeks in a month).

### Transformation

When you import data like this, you often have to check for errors, deal with special values, convert data into different formats, and perform calculations. These operations are called **data cleaning**.

First of all, ReadFemPreg() has a function CleanFemResp() within it that cleans it. Open the module again in the text editor and edit ReadFemPreg() so that CleanFemResp() is has an input to decide whether to clean it or not. <br> such as:
>ReadFemPreg(clean=True)

After loading the unclean file again. (You might have to restart the kernel to get the new function imported) Code the following Data cleaning transformation processes:

In [15]:
preg = nsfg.ReadFemPreg(dct_file='../Think_Stats/Thinkstats2/2002FemPreg.dct',
                      dat_file='Thinkstats2/2002FemPreg.dat.gz', clean = True)
preg.describe()

caseid      pregordr   howpreg_n  howpreg_p    moscurrp  \
count  13593.000000  13593.000000  352.000000  349.00000  352.000000   
mean    6216.526595      2.349150   15.144886    1.34384    4.647727   
std     3645.417341      1.577807   13.922211    0.47567    2.527523   
min        1.000000      1.000000    0.000000    1.00000    0.000000   
25%     3022.000000      1.000000    5.000000    1.00000    2.000000   
50%     6161.000000      2.000000    9.000000    1.00000    5.000000   
75%     9423.000000      3.000000   23.000000    2.00000    7.000000   
max    12571.000000     19.000000   99.000000    2.00000    9.000000   

       nowprgdk      pregend1   pregend2     nbrnaliv    multbrth  \
count  3.000000  13241.000000  18.000000  9144.000000  163.000000   
mean   3.666667      4.650177   4.055556     1.022419    1.834356   
std    4.618802      1.849790   1.696787     0.190098    1.630208   
min    1.000000      1.000000   1.000000     1.000000    1.000000   
25%    1.000000      3.000000   3.000000     1.000000    1.000000   
50%    1.000000      6.000000   4.000000     1.000000    1.000000   
75%    5.000000      6.000000   6.000000     1.000000    1.000000   
max    9.000000      9.000000   6.000000     5.000000    5.000000   

          ...         laborfor_i    religion_i  metro_i       basewgt  \
count     ...       13593.000000  13593.000000  13593.0  13593.000000   
mean      ...           0.000809      0.003016      0.0   4216.271164   
std       ...           0.028437      0.058727      0.0   3982.680473   
min       ...           0.000000      0.000000      0.0     64.577101   
25%       ...           0.000000      0.000000      0.0   2335.445237   
50%       ...           0.000000      0.000000      0.0   3409.648504   
75%       ...           0.000000      0.000000      0.0   4869.941451   
max       ...           1.000000      2.000000      0.0  99707.832014   

       adj_mod_basewgt       finalwgt        secu_p          sest  cmintvw  \
count     13593.000000   13593.000000  13593.000000  13593.000000      0.0   
mean       5383.982581    8196.422280      1.487310     44.083352      NaN   
std        5640.499431    9325.918114      0.499857     24.110403      NaN   
min          71.201194     118.656790      1.000000      1.000000      NaN   
25%        2798.048902    3841.375308      1.000000     25.000000      NaN   
50%        4127.220642    6256.592133      1.000000     45.000000      NaN   
75%        5795.692880    9432.360931      2.000000     65.000000      NaN   
max      157143.686687  261879.953864      2.000000     84.000000      NaN   

       totalwgt_lb  
count  9038.000000  
mean      7.265628  
std       1.408293  
min       0.125000  
25%       6.500000  
50%       7.375000  
75%       8.125000  
max      15.437500  

[8 rows x 244 columns]

Now import the unclean dataframe and do the following transformations:

agepreg contains the mother’s age at the end of the pregnancy. In the data file, agepreg is encoded as an integer number of centiyears. So first divide each element of agepreg by 100, yielding a floating-point value in
years.

In [16]:
preg = nsfg.ReadFemPreg(dct_file='../Think_Stats/Thinkstats2/2002FemPreg.dct',
                      dat_file='Thinkstats2/2002FemPreg.dat.gz')

preg['agepreg'] = preg['agepreg']/100
preg.agepreg.head()

0    0.3316
1    0.3925
2    0.1433
3    0.1783
4    0.1833
Name: agepreg, dtype: float64

birthwgt_lb and birthwgt_oz contain the weight of the baby, in pounds and ounces, for pregnancies that end in live birth. In addition it uses several special codes:<br/>
97 NOT ASCERTAINED<br/>
98 REFUSED<br/>
99 DON'T KNOW<br/>

1. Replace those values with nan. 

Special values encoded as numbers are dangerous because if they are not handled properly, they can generate bogus results, like a 99-pound baby. The replace method replaces these values with np.nan, a special floating- point value that represents “not a number.” The inplace flag tells replace to modify the existing Series rather than create a new one.<br/>

In [17]:
preg.loc[:,['birthwgt_lb','birthwgt_oz']] = preg.loc[:,['birthwgt_lb','birthwgt_oz']].replace([97,98,99])
preg.loc[:,['birthwgt_lb','birthwgt_oz']][13:18]

birthwgt_lb  birthwgt_oz
13          NaN          NaN
14          NaN          NaN
15          7.0         11.0
16          7.0          8.0
17          6.0          5.0

Be careful nan is not a string is a numpy object. np.nan

As part of the IEEE floating-point standard, all mathematical operations
return nan if either argument is nan:<br/>
```
>import numpy as np
>np.nan / 100.0
nan
```
So computations with nan tend to do the right thing, and most pandas functions handle nan appropriately. But dealing with missing data will be a recurring issue.

Create a new column totalwgt_lb that com- bines pounds and ounces into a single quantity, in pounds.<br>
One important note: when you add a new column to a DataFrame, you must use dictionary syntax

In [18]:
preg['totalwgt_lb'] = preg['birthwgt_lb'] + preg['birthwgt_oz']*0.0625
preg['totalwgt_lb'].head(20)

0     8.8125
1     7.8750
2     9.1250
3     7.0000
4     6.1875
5     8.5625
6     9.5625
7     8.3750
8     7.5625
9     6.6250
10    7.8125
11    7.0000
12    4.0000
13       NaN
14       NaN
15    7.6875
16    7.5000
17    6.3125
18       NaN
19    8.7500
Name: totalwgt_lb, dtype: float64

Compare them with the results the fuction when it cleans the data

In [19]:
preg = nsfg.ReadFemPreg(dct_file='../Think_Stats/Thinkstats2/2002FemPreg.dct',
                      dat_file='Thinkstats2/2002FemPreg.dat.gz', clean = True)

preg['totalwgt_lb_clean'] = preg['birthwgt_lb'] + preg['birthwgt_oz']*0.0625
preg['totalwgt_lb_clean'].head(20)

0     8.8125
1     7.8750
2     9.1250
3     7.0000
4     6.1875
5     8.5625
6     9.5625
7     8.3750
8     7.5625
9     6.6250
10    7.8125
11    7.0000
12    4.0000
13       NaN
14       NaN
15    7.6875
16    7.5000
17    6.3125
18       NaN
19    8.7500
Name: totalwgt_lb_clean, dtype: float64

### Validation

When data is exported from one software environment and imported into another, errors might be introduced. And when you are getting familiar with a new dataset, you might interpret data incorrectly or introduce other
misunderstandings. If you take time to validate the data, you can save time later and avoid errors. <br>
One way to validate data is to compute basic statistics and compare them with published results. For example, the NSFG codebook includes tables that summarize each variable. Here is the table for outcome, which encodes the outcome of each pregnancy:<br>

![alt text](notebookpics/number_rows_table.png "Title")

The Series class provides a method, value_counts, that counts the number of times each value appears. If we select the outcome Series from the DataFrame.<br>
Use value_counts to compare with the published data:

In [20]:
preg = nsfg.ReadFemPreg(dct_file='../Think_Stats/Thinkstats2/2002FemPreg.dct',
                      dat_file='Thinkstats2/2002FemPreg.dat.gz', clean=True)

In [21]:
preg.outcome.value_counts()

1    9148
4    1921
2    1862
6     352
5     190
3     120
Name: outcome, dtype: int64

Similarly, here is the published table for birthwgt_lb. Is there anything weird? If so, fix it.<br>
![alt text](notebookpics/number_rows_table2.png "Title")

In [22]:
preg = nsfg.ReadFemPreg(dct_file='../Think_Stats/Thinkstats2/2002FemPreg.dct',
                      dat_file='Thinkstats2/2002FemPreg.dat.gz')

In [23]:
# Solution 1
bins = pd.IntervalIndex.from_tuples([(-1, 5), (5, 6), (6, 7), (7, 8), (8, 950)])
print(pd.cut(preg['birthwgt_lb'], bins).value_counts(dropna = False).sort_index())

NaN         4509
(-1, 5]     1125
(5, 6]      2223
(6, 7]      3049
(7, 8]      1889
(8, 950]     798
Name: birthwgt_lb, dtype: int64


In [24]:
# Solution 2
sorted_index = preg.birthwgt_lb.value_counts(dropna = False).sort_index()

under6lb = sum(sorted_index[0:5])
lb6 = sorted_index[6]
lb7 = sorted_index[7]
lb8 = sorted_index[8]
over9lb = sum(sorted_index[9:99]) # More than index 15 says error... and also -1 in nan 
inapplicable = 4449 # sorted_index[10]

print('INAPPLICABLE |', inapplicable)
print('UNDER 6 POUNDS |', under6lb)
print('6 POUNDS |', lb6)
print('7 POUNDS |', lb7)
print('8 POUNDS |', lb8)
print('9 POUNDS OR MORE |', over9lb)

KeyError: 99.0

In [ ]:
# Solution 3
under6lb = sum(sorted_index[0:5])
lb6 = sorted_index[6]
lb7 = sorted_index[7]
lb8 = sorted_index[8]
over9lb = sum(sorted_index[9:99]) # More than index 15 says error... and also -1 in nan 
inapplicable = 4449 # sorted_index[10]

print('INAPPLICABLE |', inapplicable)
print('UNDER 6 POUNDS |', under6lb)
print('6 POUNDS |', lb6)
print('7 POUNDS |', lb7)
print('8 POUNDS |', lb8)
print('9 POUNDS OR MORE |', over9lb)

In [ ]:
preg = preg.loc[preg['birthwgt_lb'] <= 15.0] # More than 15 lbs doesn't make sense
preg.birthwgt_lb.value_counts(dropna = False).sort_index()

### Interpretation

To work with data effectively, you have to think on two levels at the same time: the level of statistics and the level of context.<br>
As an example, let’s look at the sequence of outcomes for a respondents.
Because of the way the data files are organized, we have to do some processing to collect the pregnancy data for each respondent.

Create a dictionary that maps each caseid to all of index for the pregnancies she has been involved on:

An output as: {1:[1,1],2:[1,1,1].....}

dont use pandas dataframe functions.

In [ ]:
preg.head()

In [ ]:
list(preg.loc[preg.caseid == 1, 'outcome'])

In [ ]:
dicto = {}

for i in preg.caseid:
    dicto[i] = list(preg.loc[preg.caseid == i, 'outcome'])
    
dicto[2]

Now make it into a function

In [ ]:
def todict(preg, column1, column2):
    dicto = {}
    for i in preg[column1]:
        dicto[i] = list(preg.loc[preg.caseid == i, column2])
    
    return dicto

In [ ]:
todict(preg,'caseid','outcome')

What are all the outcomes observed for caseid = 10229 (use your calculated dictionary)

In [ ]:
todict(preg,'caseid','outcome')[10229]

The outcome code 1 indicates a live birth. Code 4 indicates a miscarriage; that is, a pregnancy that ended spontaneously, usually with no known medical cause.

Statistically this respondent is not unusual. Miscarriages are common and there are other respondents who reported as many or more.

But remembering the context, this data tells the story of a woman who was pregnant six times, each time ending in miscarriage. Her seventh and most recent pregnancy ended in a live birth. If we consider this data with empathy,
it is natural to be moved by the story it tells.

Each record in the NSFG dataset represents a person who provided honest answers to many personal and difficult questions. We can use this data to answer statistical questions about family life, reproduction, and health. At
the same time, we have an obligation to consider the people represented by the data, and to afford them respect and gratitude.<br>

## Exercises

In [ ]:
preg = nsfg.ReadFemPreg(dct_file='../Think_Stats/Thinkstats2/2002FemPreg.dct',
                      dat_file='Thinkstats2/2002FemPreg.dat.gz')
preg.head()

Select the `birthord` column, print the value counts, and compare to results published in the [codebook](http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=A&subSec=8016&srtLabel=611933)

In [ ]:
# Solution goes here
print(preg['birthord'].value_counts( dropna = False))
print('TOTAL  ', len(preg['birthord']))

# value	label	 	Total
# .	INAPPLICABLE	 	4445
# 1	1ST BIRTH	 	4413
# 2	2ND BIRTH	 	2874
# 3	3RD BIRTH	 	1234
# 4	4TH BIRTH	 	421
# 5	5TH BIRTH	 	126
# 6	6TH BIRTH	 	50
# 7	7TH BIRTH	 	20
# 8	8TH BIRTH	 	7
# 9	9TH BIRTH	 	2
# 10	10TH BIRTH	 	1
#  	Total	 	13593

We can also use `isnull` to count the number of nans.

In [ ]:
preg['birthord'].isnull().head(15)

In [ ]:
sum(preg['birthord'].isnull()) # True = 1, False = 0

Select the `prglngth` column, print the value counts, and compare to results published in the [codebook](http://www.icpsr.umich.edu/nsfg6/Controller?displayPage=labelDetails&fileCode=PREG&section=A&subSec=8016&srtLabel=611931)

In [ ]:
bins1 = pd.IntervalIndex.from_tuples([(0, 13), (14, 26), (27, 50)])
print(pd.cut(preg['prglngth'], bins1).value_counts(dropna = False).sort_index())
print('TOTAL  ', len(preg['prglngth']))

# value	label	 	Total
# 0-13	13 WEEKS OR LESS	 	3522
# 14-26	14-26 WEEKS	 	793
# 27-50	27 WEEKS OR LONGER	 	9278
#  	Total	 	13593

To compute the mean of a column, you can invoke the `mean` method on a Series.  For example, here is the mean birthweight in pounds:

In [ ]:
preg.birthwgt_lb.mean()

Create a new column named <tt>totalwgt_kg</tt> that contains birth weight in kilograms.  Compute its mean.  Remember that when you create a new column, you have to use dictionary syntax, not dot notation.

In [ ]:
preg['totalwgt_kg'] = preg['birthwgt_lb']*0.453592
print(preg.totalwgt_kg.mean())
print(preg.birthwgt_lb.mean()*0.453592)

`nsfg.py` also provides `ReadFemResp`, which reads the female respondents file and returns a `DataFrame`:

In [ ]:
resp = nsfg.ReadFemResp(dct_file='../Think_Stats/Thinkstats2/2002FemResp.dct',
                      dat_file='Thinkstats2/2002FemResp.dat.gz')

`DataFrame` provides a method `head` that displays the first five rows:

In [ ]:
resp.head()

Select the `age_r` column from `resp` and print the value counts.  How old are the youngest and oldest respondents?

In [ ]:
counts = resp.age_r.value_counts(dropna = False).sort_index()
print(counts)
print()
print('Youngest:', counts.index[0])
print('Oldest:', counts.index[-1])

We can use the `caseid` to match up rows from `resp` and `preg`.  For example, we can select the row from `resp` for `caseid` 2298 like this:

In [ ]:
resp[resp['caseid'] == 2298]

And we can get the corresponding rows from `preg` like this:

In [ ]:
preg[preg['caseid'] == 2298]

How old is the respondent with `caseid` 1?

In [ ]:
print('Age:',int(resp.loc[resp['caseid'] == 1, 'age_r']))

What are the pregnancy lengths for the respondent with `caseid` 2298?

In [ ]:
print('Pregnancy lengths (weeks):', list(preg.loc[preg['caseid'] == 2298, 'prglngth']))

What was the birthweight of the first baby born to the respondent with `caseid` 2732?

In [ ]:
print('Birthweight:', float(preg.loc[(preg['caseid'] == 2732 )& (preg['pregordr'] == 1), 'finalwgt']))

In the repository you downloaded, you should find a file named chap01ex.py; using this file as a starting place, write a function that reads the respondent file, 2002FemResp.dat.gz.

The variable pregnum is a recode that indicates how many times each respondent has been pregnant. Print the value counts for this variable and compare them to the published results in the NSFG codebook. You can also crossvalidate the respondent and pregnancy files by comparing pregnum for each respondent with the number of records in the pregnancy file.

You can use nsfg.MakePregMap to make a dictionary that maps from each caseid to a list of indices into the pregnancy DataFrame.

## Extra exercise

The best way to learn about statistics is to work on a project you are interested in. Is there a question like, “Do first babies arrive late,” that you want to investigate?

Think about questions you find personally interesting, or items of conventional wisdom, or controversial topics, or questions that have political consequences, and see if you can formulate a question that lends itself to statistical inquiry.

Look for data to help you address the question. Governments are good sources because data from public research is often freely available. Good places to start include http://www.data.gov/, and http://www.science.gov/, and in the United Kingdom, http://data.gov.uk/.

Two of my favorite data sets are the [General Social Survey](http://www3.norc.org/gss+website/), and the [European Social Survey](http://www.europeansocialsurvey.org/).

You can also use typical Data science datasets from places like https://www.kdnuggets.com/datasets/index.html or https://www.kaggle.com/datasets

If it seems like someone has already answered your question, look closely to see whether the answer is justified. There might be flaws in the data or the analysis that make the conclusion unreliable. In that case you could perform a different analysis of the same data, or look for a better source of data.

If you find a published paper that addresses your question, you should be able to get the raw data. Many authors make their data available on the web, but for sensitive data you might have to write to the authors, provide information about how you plan to use the data, or agree to certain terms of use. Be persistent!

The best way to learn about statistics is to work on a project you are interested in. Is there a question like, “Do first babies arrive late,” that you want to investigate?

Think about questions you find personally interesting, or items of conventional wisdom, or controversial topics, or questions that have political consequences, and see if you can formulate a question that lends itself to statistical inquiry.

Look for data to help you address the question. Governments are good sources because data from public research is often freely available. Good places to start include http://www.data.gov/, and http://www.science.
gov/, and in the United Kingdom, http://data.gov.uk/.

Two of my favorite data sets are the General Social Survey at http://www3.norc.org/gss+website/, and the European Social Survey at http://www.europeansocialsurvey.org/.

You can also use typical Data science datasets from places like https://www.kdnuggets.com/datasets/index.html or https://www.kaggle.com/datasets

More:
- https://www.data.gov/
- https://www.reddit.com/r/datasets
- https://github.com/awesomedata/awesome-public-datasets
- https://fivethirtyeight.com/

If it seems like someone has already answered your question, look closely to see whether the answer is justified. There might be flaws in the data or the analysis that make the conclusion unreliable. In that case you could perform a different analysis of the same data, or look for a better source of data.

If you find a published paper that addresses your question, you should be able to get the raw data. Many authors make their data available on the web, but for sensitive data you might have to write to the authors, provide information about how you plan to use the data, or agree to certain terms of use. Be persistent!

###### Javier Fernández Suárez